snackpub/src/index.ts

```diff
@@ -0,0 +1,111 @@
+import { Server } from 'socket.io';
+
+import Env from './Env';
+import { maybeBindRedisAdapterAsync, maybeCloseRedisConnectionsAsync } from './RedisAdapter';
+import type {
+  ClientToServerEvents,
+  InterServerEvents,
+  ServerToClientEvents,
+  ShutdownSignal,
+  SocketData,
+} from './types';
+
+const debug = require('debug')('snackpub');
+
+function registerShutdownHandlers(server: Server) {
+  const shutdown = async (signal: ShutdownSignal) => {
+    console.log(
+      `Received ${signal}; the HTTP server is shutting down and draining existing connections`
+    );
+    await closeServerAsync(server);
+    await maybeCloseRedisConnectionsAsync();
+  };
+
+  // TODO: In Node 9, the signal is passed as the first argument to the listener
+  process.once('SIGHUP', () => shutdown('SIGHUP'));
+  process.once('SIGINT', () => shutdown('SIGINT'));
+  process.once('SIGTERM', () => shutdown('SIGTERM'));
+  // Nodemon sends SIGUSR2 when it restarts the process it's monitoring
+  process.once('SIGUSR2', () => shutdown('SIGUSR2'));
+}
+
+function closeServerAsync(server: Server): Promise<void> {
+  return new Promise((resolve) => {
+    server.close((error) => {
+      if (error) {
+        console.log('Failed to close server', error);
+      }
+      resolve();
+    });
+  });
+}
+
+async function runAsync() {
+  const io = new Server<ClientToServerEvents, ServerToClientEvents, InterServerEvents, SocketData>({
+    serveClient: false,
+  });
+
+  await maybeBindRedisAdapterAsync(io);
+
+  io.on('connection', (socket) => {
+    debug('onconnect', socket.handshake.address);
+    socket.on('message', (data) => {
+      // debug('onMessage', data);
+      const { channel, message, sender } = data;
+      socket.to(channel).emit('message', { channel, message, sender });
```

Comment:
Are channel names unguessable, similar to a session secret? We want to make sure that the only clients who can connect to a channel are the ones that already know about the channel.

Reply:
this is [the current implementation](https://github.com/expo/snack/blob/bd1394bda92c5a45791de13d600707dbf6081bc6/packages/snack-sdk/src/utils.ts#L10-L28) to generate channel. as mentioned on slack, as we have channel rate limiter now. i thought we could enhance the security later.

the channel authentication briefly in my thought is between snack-sdk <-> runtime. it has no difference between snackpub or pubnub.

Reply:
Ok, 60 bits should be secure enough because we have the rate limiter and Snack channels do not live long enough for someone to try billions of combinations.